STF 1.1 documentation updates (round 1) (#136)
infrawatch/documentation: doc-Service-Telemetry-Framework/modules/con_overview-of-the-servicetelemetry-object.adoc

// Do not start the title with a verb. See also _Wording of headings_
// in _The IBM Style Guide_.
[id="overview-of-the-servicetelemetry-object"]
== Overview of the ServiceTelemetry object

IMPORTANT: Versions of Service Telemetry Operator prior to v1.1.0 used a flat API (`servicetelemetry.infra.watch/v1alpha1`) interface for creating the `ServiceTelemetry` object.
In Service Telemetry Operator v1.1.0, there is a dictionary-based API interface (`servicetelemetry.infra.watch/v1beta1`) to allow for better control of {ProjectShort} deployments, including managing multi-cluster deployments natively, and allowing the management of additional storage backends in the future. Ensure that any previously created `ServiceTelemetry` objects are updated to the new interface.

To deploy the {Project}, you must create an instance of `ServiceTelemetry` in {OpenShiftShort}. The `ServiceTelemetry` object is made up of the following major configuration parameters:

* `alerting`
* `backends`
* `clouds`
* `graphing`
* `highAvailability`
* `transports`

Each of these top-level configuration parameters provides various controls for a {Project} deployment.

=== backends

> Use the `backends` parameter to control what storage backends are available for storage of metrics and events and to control the enablement of Smart Gateways, as defined by the `clouds` parameter. For more information, see xref:_clouds[].

Reviewer: This link doesn't work. I'm logging off now but will check this and rest of doc tomorrow morning!

Author: Oh it probably worked until I added an `id` to it. I think this needs to be `xref:clouds_installing-the-core-components-of-stf[]` now

> Use the `backends` parameter to control what storage backends are available for storage of metrics and events and to control the enablement of Smart Gateways, as defined by the `clouds` parameter. For more information, see xref:_clouds[].

Reviewer: ```suggestion
Use the `backends` parameter to control which storage backends are available for storage of metrics and events, and to control the enablement of Smart Gateways, as defined by the `clouds` parameter. For more information, see xref:clouds_installing-the-core-components-of-stf[].
```

The available storage backends include `metrics` and `events`. The metrics backends include Prometheus and the events backends include ElasticSearch.

==== Enabling Prometheus as a storage backend for metrics

.Procedure

* To enable Prometheus as a storage backend for metrics, configure the `ServiceTelemetry` object:
+
[source,yaml]
----
apiVersion: infra.watch/v1beta1
kind: ServiceTelemetry
metadata:
  name: default
  namespace: service-telemetry
spec:
  backends:
    metrics:
      prometheus:
        enabled: true
----

> As the parameters are a dictionary of parameters, referencing them in the documentation will to be presented using dotted notation. For example, the configuration example for enabling the Prometheus backend for storage of metrics would be noted as `backends.metrics.prometheus.enabled: true`.

Reviewer: ```suggestion

```

==== Enabling ElasticSearch as a storage backend for events

> To enable events support in {ProjectShort} requires, you must enable the Elastic Cloud for Kubernetes Operator. For more information, see xref:subscribing-to-elastic-cloud-on-kubernetes-operator[]. By default, ElasticSearch storage of events is disabled. For more information, see xref:deploying-stf-to-the-openshift-environment-without-elasticsearch[].

Reviewer: ```suggestion
To enable events support in {ProjectShort}, you must enable the Elastic Cloud for Kubernetes Operator. For more information, see xref:subscribing-to-elastic-cloud-on-kubernetes-operator[]. By default, ElasticSearch storage of events is disabled. For more information, see xref:deploying-stf-to-the-openshift-environment-without-elasticsearch[].
```

[id="clouds_{context}"]
=== clouds

> Use the `clouds` parameter to control which Smart Gateway objects are deployed, thereby providing the interface for multiple monitoring cloud environments to connect to an instance of {ProjectShort}. If a supporting backend is available, then metrics and events Smart Gateways for the default cloud configuration are created. By default the Service Telemetry Operator will create Smart Gateways for `cloud1`.

Reviewer: ```suggestion
Use the `clouds` parameter to control which Smart Gateway objects are deployed, thereby providing the interface for multiple monitoring cloud environments to connect to an instance of {ProjectShort}. If a supporting backend is available, then metrics and events Smart Gateways for the default cloud configuration are created. By default, the Service Telemetry Operator creates Smart Gateways for `cloud1`.
```

> You can create a list of cloud objects to control which Smart Gateways are created for each cloud defined. Each cloud is made up of data types and collectors. Data types are `metrics` or `events`. Each data type is made up of a list of collectors and the message bus subscription address. Available collectors are `collectd` and `ceilometer`. Ensure that the subscription address for each of these collectors is unique for every cloud, data type, and collector combination.

Reviewer: ```suggestion
You can create a list of cloud objects to control which Smart Gateways are created for each defined cloud. Each cloud consists of data types. Data types can be `metrics` or `events`. Each data type is made up of a list of collectors and the message bus subscription address. Available collectors are `collectd` and `ceilometer`. Ensure that the subscription address for each of these collectors is unique for every cloud, data type, and collector combination.
```

The default `cloud1` configuration is represented by the following `ServiceTelemetry` object, providing subscriptions and data storage of metrics and events for both collectd and Ceilometer data collectors for a particular cloud instance:

[source,yaml]
----
apiVersion: infra.watch/v1beta1
kind: ServiceTelemetry
metadata:
  name: stf-default
  namespace: service-telemetry
spec:
  clouds:
    - name: cloud1
      metrics:
        collectors:
          - collectorType: collectd
            subscriptionAddress: collectd/telemetry
          - collectorType: ceilometer
            subscriptionAddress: anycast/ceilometer/metering.sample
      events:
        collectors:
          - collectorType: collectd
            subscriptionAddress: collectd/notify
          - collectorType: ceilometer
            subscriptionAddress: anycast/ceilometer/event.sample
----

> Each item of the `clouds` parameter represents a cloud instance. The cloud instances are made up of 3 top-level parameters: `name`, `metrics`, and `events`. The metrics and events parameters represent the corresponding backend for storage of that data type. The `collectors` parameter then specifies a list of objects made up of two parameters, `collectorType` and `subscriptionAddress`, and these represent an instance of the Smart Gateway. The collectorType specifies data collected by either collectd or Ceilometer. The subscriptionAddress parameter provides the {MessageBus} address that a Smart Gateway instance should subscribe to.

Reviewer: ```suggestion
Each item of the `clouds` parameter represents a cloud instance. The cloud instances are made up of three top-level parameters: `name`, `metrics`, and `events`. The `metrics` and `events` parameters represent the corresponding backend for storage of that data type. The `collectors` parameter specifies a list of objects made up of two parameters, `collectorType` and `subscriptionAddress`, and these represent an instance of the Smart Gateway. The `collectorType` parameter specifies data collected by either collectd or Ceilometer. The `subscriptionAddress` parameter provides the {MessageBus} address to which a Smart Gateway instance subscribes.
```

=== alerting

> Use the `alerting` parameter to control creation of an link:https://prometheus.io/docs/alerting/latest/alertmanager/[Alertmanager] instance and the configuration of the storage backend. By default, alerting is enabled. For more information, see xref:alerts_advanced-features[].

Reviewer: ```suggestion
Use the `alerting` parameter to control creation of an link:https://prometheus.io/docs/alerting/latest/alertmanager/[Alertmanager] instance and the configuration of the storage backend. By default, `alerting` is enabled. For more information, see xref:alerts_advanced-features[].
```

=== graphing

> Use the `graphing` parameter to control the creation of a link:https://grafana.com/docs/grafana/latest/getting-started/what-is-grafana/[Grafana] instance. By default, graphing is disabled. For more information, see xref:dashboards_advanced-features[].

Reviewer: ```suggestion
Use the `graphing` parameter to control the creation of a link:https://grafana.com/docs/grafana/latest/getting-started/what-is-grafana/[Grafana] instance. By default, `graphing` is disabled. For more information, see xref:dashboards_advanced-features[].
```

=== highAvailability

> Use The `highAvailability` parameter to control the instantiation of multiple copies of {ProjectShort} components to reduce recovery time of components that fail or are rescheduled. By default highAvailability is disabled. For more information, see xref:high-availability_advanced-features[].

Reviewer: ```suggestion
Use The `highAvailability` parameter to control the instantiation of multiple copies of {ProjectShort} components to reduce recovery time of components that fail or are rescheduled. By default, `highAvailability` is disabled. For more information, see xref:high-availability_advanced-features[].
```
